src/main/java/net/ripe/rpki/rsyncit/rrdp/RrdpFetcher.java

```diff
@@ -162,18 +179,30 @@ private static void validateSnapshotStructure(int notificationSerial, String sna
         }
     }
 
-    private ProcessPublishElementResult processPublishElements(Element doc) throws XPathExpressionException {
+    private ProcessPublishElementResult processPublishElements(Element doc, Instant lastModified) throws XPathExpressionException {
         var queryPublish = XPathFactory.newDefaultInstance().newXPath().compile("/snapshot/publish");
         final NodeList publishedObjects = (NodeList) queryPublish.evaluate(doc, XPathConstants.NODESET);
 
+        // Generate timestamp that will be tracked per object and used as FS modification timestamp.
+        // Use last-modified header from the snapshot if available, otherwise truncate current time
+        // to the closest hour -- it is unlikely that different instances will have clocks off by a lot,
+        // so rounding down to an hour should generate the same timestamps _most of the time_.
+        //
+        var defaultTimestamp = lastModified != null ? lastModified : Instant.now().truncatedTo(ChronoUnit.HOURS);
+
+        // This timestamp is only needed for marking objects in the timestamp cache.
         var now = Instant.now();
-        var collisionCount = new AtomicInteger();
 
+        var collisionCount = new AtomicInteger();
         var decoder = Base64.getDecoder();
 
-        var t = Time.timed(() -> IntStream
+        var objectItems = IntStream
             .range(0, publishedObjects.getLength())
             .mapToObj(publishedObjects::item)
+            .toList();
+
+        var t = Time.timed(() -> objectItems
```

Comment:
We could use https://micrometer.io/docs/concepts#_recording_blocks_of_code and get a metric - but no logging - for free :)

Reply:
It would make sense to re-iterated on that indeed, this `Time.timed` is annoying when it comes to checked exceptions. But I would do it in a separate PR.